### PR TITLE
[example]add bucketing/batchnorm and improved performance for speech_recognition example

### PR DESCRIPTION
**Pull request by Soonhwan-Kwon**
add bucketing and batch norm scheme for speech_recognition example to improve training performance
add flexibility to support various optimizer
fix bug in generating bi-graphemes dictionray
add option not to save csv file to improve performane and reduce disk space requirement
and it solves performance issue(speed) related to #6418
and may related to issues #3076, #6115, #2663
when using both bucketing(variable length) and batch norm
although it is not a solution for time-step wise but layer wise approach for batch normalization

### CODE REVIEW

**Comment by Soonhwan-Kwon** [comment]
@piiswrong I tested on latest build and it works well. Please let me know if there is something to do.

**Comment by piiswrong** [comment]
Thanks. LGTM
